### PR TITLE
Node recovery from stalled Stateproof chain

### PR DESCRIPTION
**Pull request by algoidan**
## Summary
In case state proof chain is being delayed nodes have the ability to recover and "catch-up" on not-yet-confirmed state proofs. In order to do so, there are some resources allocated for that target.

Nodes store the following resources and remove them only when a state proof is confirmed on-chain
1 -Voters array, participation tree, and proven weight
2 -builder's data for each state proof round
3-signatures on state proof messages

If state proof chain is delayed too much those resources will never be released and might lead to extensive memory consumption.

### CODE REVIEW

File: stateproof/builder.go [hunk 1]
```diff
@@ -327,19 +329,45 @@ func (spw *Worker) broadcastSigs(brnd basics.Round, proto config.ConsensusParams
...
-func (spw *Worker) deleteOldSigs(nextStateProof basics.Round) {
+func lowestRoundToRemove(currentHdr bookkeeping.BlockHeader) basics.Round {
+	proto := config.Consensus[currentHdr.CurrentProtocol]
+	nextStateProofRnd := currentHdr.StateProofTracking[protocol.StateProofBasic].StateProofNextRound
+	if proto.StateProofInterval == 0 {
+		return nextStateProofRnd
+	}
+
+	recentRoundOnRecoveryPeriod := basics.Round(uint64(currentHdr.Round) - uint64(currentHdr.Round)%proto.StateProofInterval)
+	oldestRoundOnRecoveryPeriod := recentRoundOnRecoveryPeriod.SubSaturate(basics.Round(proto.StateProofInterval * proto.StateProofRecoveryInterval))
```
Comment (algonautshant) [inline stateproof/builder.go:340]: Here, shouldn't the protocol used be the one from round `recentRoundOnRecoveryPeriod` instead of `proto`, which is the round of `currentHdr`?

**Review by algonautshant — COMMENTED** [review]
Can you please update the PR title to be more specific? 
It is not "Stateproof recovery", it is "Node recovery from stalled Stateproof chain". 
The title is implying that the StateProof is recovered, while it is not.

File: test/e2e-go/features/stateproofs/stateproofs_test.go [hunk 2]
```diff
@@ -173,3 +175,148 @@ func verifyStateProofForRound(r *require.Assertions, libgoal libgoal.Client, res
...
+		// send a dummy payment transaction.
+		sendPayment(r, &fixture, rnd)
+
+		err = fixture.WaitForRound(rnd, 30*time.Second)
+		r.NoError(err)
+
+		blk, err := libgoal.BookkeepingBlock(rnd)
+		r.NoErrorf(err, "failed to retrieve block from algod on round %d", rnd)
+
+		//t.Logf("Round %d, block %v\n", rnd, blk)
```
[the remark this answers is not in the document]
Comment (algoidan) [reply test/e2e-go/features/stateproofs/stateproofs_test.go:304]: Maybe I've missed something, what resource do I need to clean?

File: stateproof/builder.go (hunk 1 above)
Comment (algoidan) [reply stateproof/builder.go:340]: from my understanding, 
in the current implementation, if we choose to change the `StateProofRecoveryInterval` it will take effect immediately. if we change the code and take  `StateProofRecoveryInterval` from `recentRoundOnRecoveryPeriod` , changing `StateProofRecoveryInterval` might take effect one interval after.

Do you think I got it right?
if so, would it be simpler to keep it as is and a change `StateProofRecoveryInterval` will take effect immediately?

File: config/consensus.go [hunk 3]
```diff
@@ -387,6 +387,11 @@ type ConsensusParams struct {
 	// StateProofStrengthTarget represents either k+q (for pre-quantum security) or k+2q (for post-quantum security)
 	StateProofStrengthTarget uint64
 
+	// StateProofRecoveryInterval represents the number of state proof intervals that the network will try to catch-up with.
+	// When the difference between the latest state proof and the current round will be greater than value, Nodes will
+	// release resources allocated for creating state proofs.
+	StateProofRecoveryInterval uint64
```
[the remark this answers is not in the document]
Comment (algoidan) [reply config/consensus.go:393]: I think this should be a consensus param since different values would on different node might cause nodes to disagree on stateproof transactions (they might remove data used for the verification)

[the remark this answers is not in the document]
Comment (algoidan) [reply config/consensus.go:393]: @gmalouf @cce, We would like to hear your thoughts on this.

[the remark this answers is not in the document]
Comment (algonautshant) [reply config/consensus.go:393]: Here is why I think this parameter should be a configuration parameter, and not a consensus parameter:

This parameter decides how much memory a system can tolerate. This parameter needs to be set to a reasonably large value, to provide some tolerance in case the SP is delayed, and perhaps give enough time for intervention when something wrong happens.

However, a machine with limited amount of memory will crash when the memory usage goes up because of this parameter.

As a more concrete example, let's say there are two machines, one with 95% stake and a lot of memory, and another machine, with 5% stake, and limited memory. It is not a good idea to have the same value for this parameter to all of these machines. 

The machines with high stake and high memory, can and should set a large value for this parameter, while others, with limited memory, can only tolerate a small value. 

Therefore, I think this parameter should not be a consensus parameter. 

@gmalouf @brianolson your thoughts.

[the remark this answers is not in the document]
Comment (brianolson) [reply config/consensus.go:393]: How big is a state proof? (serialized? in memory?)
I'm not sure I understand the work flow, but if something asks `algod` to validate against state proofs, does it return three answers {valid, invalid, I don't know} ? If "I don't know" is an acceptable answer, then better nodes hold more, cheaper nodes hold fewer.

[the remark this answers is not in the document]
Comment (algoidan) [reply config/consensus.go:393]: a key observation in our discussion is that there are two parts of data needed for stateproof
1 - data needed to create a stateproof transaction (participation array, bound of signatures etc)
2- data needed to verify a stateproof transaction on-chain. (need to block header from 256 rounds before the stateproof)

and currently, the new parameter affects both parts 
In my opinion, regarding the second part - all node must be synced on what blocks they have in order to verify a stateproof transaction. Having different parameter values might lead to disagreements.

However, regarding the first part, we can definitely extend this and allow some beefy servers to store more data (this will be implemented using a configuration parameter and not consensus ). This is part of what we plan to implement next quarter.

really hope this addresses your concern

[the remark this answers is not in the document]
Comment (gmalouf) [reply config/consensus.go:393]: Is it really only relay nodes that have to worry about this? If so, we should ensure their minimum recommended specs can hold up over some delayed period of reason.

[the remark this answers is not in the document]
Comment (algonautshant) [reply config/consensus.go:393]: No. All nodes that have online accounts. This parameter will impact their ability to sign blocks with SP transaction in them.

[the remark this answers is not in the document]
Comment (algonautshant) [reply config/consensus.go:393]: Yes, we need a live discussion for this, absolutely. And the value of 10, which is too much for Brian :-), is of very limited use if set that low. 
Also, once added a consensus parameter, it will be there forever!

[the remark this answers is not in the document]
Comment (cce) [reply config/consensus.go:393]: @algoidan says:
> > a key observation in our discussion is that there are two parts of data needed for stateproof
> > 1 - data needed to create a stateproof transaction (participation array, bound of signatures etc)
> > 2- data needed to verify a stateproof transaction on-chain. (need to block header from 256 rounds before the stateproof)
> >
> > and currently, the new parameter affects both parts
> > In my opinion, regarding the second part - all nodes must be synced on what blocks they have in order to verify a stateproof transaction. Having different parameter values might lead to disagreements.

@algonautshant replies:
> This is the argument for having it as a consensus parameter. But I would say that, accounts will either vote yes, or no to the transaction/block. If enough "yes" votes are obtained, then all is good. The idea that all nodes must be "synced", is not going to happen.
>
> In the event of a delayed SP, either all will say "No" (parameter too small), or the weak ones will crash, and others will say "Yes" (parameter big enough).
>
> No. All nodes that have online accounts. This parameter will impact their ability to sign blocks with SP transaction in them.

I agree with Idan on #2. There cannot be disagreement over whether a transaction is valid. All nodes (whether relay, participating, non-participating) must be able to validate transactions and block proposals. I disagree with "If enough 'yes' votes are obtained, then all is good. The idea that all nodes must be 'synced', is not going to happen."

My understanding is that consensus means every node can validate all transactions, proposals, votes, and blocks for themselves. Relaxing this requirement introduces subjectivity into whether block proposals or transactions are valid, and would lead to consensus failing to find agreement (and stalling) if enough nodes disagree on the validity of the same transactions/blocks.

@algoidan goes on:
> However, regarding the first part, we can definitely extend this and allow some beefy servers to store more data (this will be implemented using a configuration parameter and not consensus ). This is part of what we plan to implement next quarter.

Would it help to give this parameter a different name perhaps?

[the remark this answers is not in the document]
Comment (brianolson) [reply config/consensus.go:393]: Talking this over with Shant just now, `StateProofInterval` and `StateProofRecoveryInterval` _must_ be consensus parameters because they govern evaluating new transactions in a block (the recording of a state proof is a new txn type) and everyone must have the same rules for evaluating that txn. *Also* this commits us to some extra storage on each node (at least (32 byte addr)*(1000 top accounts)*(10 state proof periods) == 320,000 bytes) to store which accounts were valid signers for each state proof we may yet need to check signatures for. I was going to say 'and we should have a discussion about whether to commit that storage', but 320kB ought to be fine.

**Comment by gmalouf** [comment]
Sounds like you have your answer Idan :)

File: test/e2e-go/features/stateproofs/stateproofs_test.go [hunk 4]
```diff
@@ -173,3 +175,154 @@ func verifyStateProofForRound(r *require.Assertions, libgoal libgoal.Client, res
 	r.NoError(err)
 	return stateProofMessage, nextStateProofBlock
 }
+
+func TestStateProofsRecovery(t *testing.T) {
```
Comment (algonautshant) [inline test/e2e-go/features/stateproofs/stateproofs_test.go:179]: It is not clear what this test is about. Some comments will be appreciated here.

File: test/e2e-go/features/stateproofs/stateproofs_test.go [hunk 5]
```diff
@@ -173,3 +175,154 @@ func verifyStateProofForRound(r *require.Assertions, libgoal libgoal.Client, res
...
+			// that has a merkle root.
+			if lastStateProofBlock.Round() == 0 {
+				lastStateProofBlock = blk
+			}
+		}
+
+		for lastStateProofBlock.Round() != 0 && lastStateProofBlock.Round()+basics.Round(consensusParams.StateProofInterval) < blk.StateProofTracking[protocol.StateProofBasic].StateProofNextRound {
+			nextStateProofRound := uint64(lastStateProofBlock.Round()) + consensusParams.StateProofInterval
+
+			t.Logf("found a state proof for round %d at round %d", nextStateProofRound, blk.Round())
```
[the remark this answers is not in the document]
Comment (algoidan) [reply test/e2e-go/features/stateproofs/stateproofs_test.go:250]: This test is a bit long
I think it is a good idea to show progress.

File: ledger/voters.go [hunk 6]
```diff
@@ -186,14 +179,39 @@ func (vt *votersTracker) newBlock(hdr bookkeeping.BlockHeader) {
...
+// notice that if state proof chain is delayed, votersForRoundCache will not be larger than
+// StateProofRecoveryInterval + 1
+// ( In order to be able to build and verify X stateproofs back we need X + 1 voters data )
+func (vt *votersTracker) removeOldVoters(hdr bookkeeping.BlockHeader) {
+	// we calculate the lowest round for recovery according to the newest round (might be different from the rounds on cache)
+	proto := config.Consensus[hdr.CurrentProtocol]
+	recentRoundOnRecoveryPeriod := basics.Round(uint64(hdr.Round) - uint64(hdr.Round)%proto.StateProofInterval)
+	oldestRoundOnRecoveryPeriod := recentRoundOnRecoveryPeriod.SubSaturate(basics.Round(proto.StateProofInterval * proto.StateProofRecoveryInterval))
+
+	for r, tr := range vt.votersForRoundCache {
```
[the remark this answers is not in the document]
Comment (algoidan) [reply ledger/voters.go:195]: Basically, You are right, and the logic you've suggested will work fine.
However, I think that the current approach is simpler (we don't need a store state for example- and we might have more edge cases).
Since the map is quite small (it usually contains 0-3 elements and in the worst case contains 10 elements). I suggest we stick to a simpler code.

File: test/e2e-go/features/stateproofs/stateproofs_test.go [hunk 7]
```diff
@@ -173,3 +176,169 @@ func verifyStateProofForRound(r *require.Assertions, libgoal libgoal.Client, res
...
+	nc := nodecontrol.MakeNodeController(fixture.GetBinDir(), dir)
+	nc.FullStop()
+
+	restClient, err := fixture.NC.AlgodClient()
+	r.NoError(err)
+
+	var lastStateProofBlock bookkeeping.Block
+	var lastStateProofMessage stateproofmsg.Message
+	libgoal := fixture.LibGoalClient
+	for rnd := uint64(2); rnd <= consensusParams.StateProofInterval*(expectedNumberOfStateProofs+1); rnd++ {
```
Comment (algonautshant) [inline test/e2e-go/features/stateproofs/stateproofs_test.go:230]: This loop condition is problematic. If the SP is delayed, which is normal, the test will fail. 
It is better to set some upper bound, and break the loop when the desired number of SPs are observed. 
```suggestion
	for rnd := uint64(2); rnd <= consensusParams.StateProofInterval*(expectedNumberOfStateProofs+10); rnd++ {
```

File: test/e2e-go/features/stateproofs/stateproofs_test.go [hunk 8]
```diff
@@ -173,3 +176,169 @@ func verifyStateProofForRound(r *require.Assertions, libgoal libgoal.Client, res
...
+		for lastStateProofBlock.Round() != 0 && lastStateProofBlock.Round()+basics.Round(consensusParams.StateProofInterval) < blk.StateProofTracking[protocol.StateProofBasic].StateProofNextRound {
+			nextStateProofRound := uint64(lastStateProofBlock.Round()) + consensusParams.StateProofInterval
+
+			t.Logf("found a state proof for round %d at round %d", nextStateProofRound, blk.Round())
+			// Find the state proof transaction
+			stateProofMessage, nextStateProofBlock := verifyStateProofForRound(r, libgoal, restClient, nextStateProofRound, lastStateProofMessage, lastStateProofBlock, consensusParams)
+			lastStateProofMessage = stateProofMessage
+			lastStateProofBlock = nextStateProofBlock
+		}
+	}
```
Comment (algonautshant) [inline test/e2e-go/features/stateproofs/stateproofs_test.go:267]:
```suggestion
		if consensusParams.StateProofInterval*expectedNumberOfStateProofs == uint64(lastStateProofBlock.Round()) {
			break
		}
	}
```

File: test/e2e-go/features/stateproofs/stateproofs_test.go [hunk 9]
```diff
@@ -173,3 +176,169 @@ func verifyStateProofForRound(r *require.Assertions, libgoal libgoal.Client, res
...
+	r.Equalf(consensusParams.StateProofInterval*expectedNumberOfStateProofs, uint64(lastStateProofBlock.Round()), "the expected last state proof block wasn't the one that was observed")
+}
+
```
Comment (algonautshant) [inline test/e2e-go/features/stateproofs/stateproofs_test.go:270]:
```suggestion

// TestUnableToRecoverFromLaggingStateProofChain tests that the network continues after it fails to create SPs before StateProofRecoveryInterval
// It stops one of the nodes to prevent the SP creation and starts it after StateProofRecoveryInterval deadline
```

File: test/e2e-go/features/stateproofs/stateproofs_test.go [hunk 10]
```diff
@@ -173,3 +176,169 @@ func verifyStateProofForRound(r *require.Assertions, libgoal libgoal.Client, res
...
+	r.NoError(err)
+
+	dir, err := fixture.GetNodeDir("Node4")
+	nc := nodecontrol.MakeNodeController(fixture.GetBinDir(), dir)
+	nc.FullStop()
+
+	var lastStateProofBlock bookkeeping.Block
+
+	libgoal := fixture.LibGoalClient
+	for rnd := uint64(2); rnd <= consensusParams.StateProofInterval*(expectedNumberOfStateProofs+1); rnd++ {
```
Comment (algonautshant) [inline test/e2e-go/features/stateproofs/stateproofs_test.go:313]: Give the network more time after restarting the node, and make sure the network continues to make progress. 
```suggestion
	for rnd := uint64(2); rnd <= consensusParams.StateProofInterval*(consensusParams.StateProofRecoveryInterval+3); rnd++ {
```

**Review by algonautshant — APPROVED** [review]
A few more comments to the e2e tests.

Looks great! Thanks for all the updates.
